app/controllers/mixins/actions/vm_actions/retire.rb

```diff
@@ -127,26 +92,44 @@ def retire
           drop_breadcrumb(:name => _("Retire %{name}") % {:name => ui_lookup(:models => kls.to_s)},
                           :url  => "/#{session[:controller]}/retire")
           session[:cat] = nil                 # Clear current category
-          @retireitems = kls.find(session[:retire_items]).sort_by(&:name) # Get the db records
           build_targets_hash(@retireitems)
           @view = get_db_view(kls)              # Instantiate the MIQ Report view object
           @view.table = MiqFilter.records2table(@retireitems, @view.cols + ['id'])
           if @retireitems.length == 1 && !@retireitems[0].retires_on.nil?
             t = @retireitems[0].retires_on                                         # Single VM, set to current time
             w = @retireitems[0].retirement_warn if @retireitems[0].retirement_warn # Single VM, get retirement warn
-          else
-            t = nil
           end
-          session[:retire_date] = t.nil? ? nil : "#{t.month}/#{t.day}/#{t.year}"
+          session[:retire_date] = "#{t.month}/#{t.day}/#{t.year}" unless t.nil?
           session[:retire_warn] = w
           @in_a_form = true
           @edit ||= {}
           @edit[:object_ids] = @retireitems
           session[:edit] = @edit
           @refresh_partial = "shared/views/retire" if @explorer || @layout == "orchestration_stack"
         end
+
+        private
+
+        def handle_form_buttons(kls)
+          if params[:button] == "cancel"
+            flash = _("Set/remove retirement date was cancelled by the user")
+            @sb[:action] = nil
+          elsif params[:button] == "save"
+            if params[:retire_date].blank?
+              flash = session[:retire_items].length == 1 ? flash = _("Retirement date removed") : _("Retirement dates removed")
```

Comment:
when touching it, please, use the correct i18n function (see `n()` in http://manageiq.org/docs/guides/i18n)